File: algorithms/js/37.解数独.js
/*
 * @lc app=leetcode.cn id=37 lang=javascript
 *
 * [37] 解数独
 */

// @lc code=start
/**
 *
 * @param {number} row
 * @param {number} col
 * @param {character[][]} board
 */
const fillExists = (row, col, board) => {
  const exists = Array.from({ length: 10 }, (_) => 0);
  for (let i = 0; i < 9; i++) {
    if (board[row][i] !== ".") exists[board[row][i]] = 1;
  }
  for (let i = 0; i < 9; i++) {
    if (board[i][col] !== ".") exists[board[i][col]] = 1;
  }
  const rowFrom = Math.floor(row / 3) * 3;
  const colFrom = Math.floor(col / 3) * 3;
  for (let i = rowFrom; i < rowFrom + 3; i++) {
    for (let j = colFrom; j < colFrom + 3; j++) {
      if (board[i][j] !== ".") exists[board[i][j]] = 1;
    }
  }
  return exists;
};
/**
 *
 * @param {character[][]} board
 * @param {number} level
 */
const dfs = (board, level) => {
  // terminator
  if (level === 81) return true;
  const row = Math.floor(level / 9);
  const col = level % 9;
  if (board[row][col] !== ".") return dfs(board, level + 1);
  // existed number
  const exists = fillExists(row, col, board);
  for (let i = 1; i < 10; i++) {
    if (!exists[i]) {
      board[row][col] = String(i);
      if (dfs(board, level + 1)) return true;
      board[row][col] = ".";
    }
  }
  // not found valid number
  return false;
};
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solveSudoku = function(board) {
  dfs(board, 0);
};
// @lc code=end

File: algorithms/golang/37.解数独.go
/*
 * @lc app=leetcode.cn id=37 lang=golang
 *
 * [37] 解数独
 */

// @lc code=start

/*
 * DFS, 共91个位置，从头开始对每个位置尝试1~9，并下探到新的位置，若棋盘不合法则回退并恢复状态
 */

var rows [9][9]bool
var cols [9][9]bool
var blocks [3][3][9]bool

func helper(boardPtr *[][]byte, pos int) bool {
	if pos == 81 {
		return true
	}
	board := *boardPtr
	row, col := pos/9, pos%9
	if board[row][col] == '.' {
		for n := 0; n < 9; n++ {
			if !blocks[row/3][col/3][n] && !rows[row][n] && !cols[col][n] {
				val := '1' + byte(n)
				board[row][col] = val
				blocks[row/3][col/3][n] = true
				rows[row][n] = true
				cols[col][n] = true
				if helper(boardPtr, pos+1) {
					return true
				}
				board[row][col] = '.'
				blocks[row/3][col/3][n] = false
				rows[row][n] = false
				cols[col][n] = false
			}
		}
		return false
	}
	return helper(boardPtr, pos+1)

}
func solveSudoku(board [][]byte) {
	rows = [9][9]bool{}
	cols = [9][9]bool{}
	blocks = [3][3][9]bool{}
	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			if board[i][j] != '.' {
				n := board[i][j] - '0' - 1
				rows[i][n] = true
				cols[j][n] = true
				blocks[i/3][j/3][n] = true
			}
		}
	}
	helper(&board, 0)
}

// @lc code=end
